From e67269a910b4e698a8032835d352f7a457b8e0e7 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Thu, 29 Nov 2012 17:58:13 -0700
Subject: [PATCH] Genericize Project Dashboard Types.

Change the Project Dashboard internals and externals to
have a dashboard type.  Use the new dashboard type to
represent "DEFAULT" dashboards instead of the original
custom default mechanism.  This new type mechanism will
make it possible to define other types of dashboards in
the future (say a CHANGE dashboard), but this change does
not attempt to do so yet.  The only visible effect that
this change should have, is that DEFAULT dashboards are
now displayed on the ProjectDashboardsScreen with a D
instead of a green checkbox.  This makes it possible to
later add more types and disambiguate between them.

Change-Id: Id4b4580b32fbb95e13a3e48122d61808aa9bef08
---
 .../java/com/google/gerrit/client/Dispatcher.java  |    7 +-
 .../client/dashboards/DashboardConstants.java      |    1 -
 .../dashboards/DashboardConstants.properties       |    1 -
 .../gerrit/client/dashboards/DashboardInfo.java    |   20 +++-
 .../gerrit/client/dashboards/DashboardList.java    |    4 +-
 .../gerrit/client/dashboards/DashboardsTable.java  |   22 ++-
 .../rpc/changedetail/ChangeDetailFactory.java      |    4 +
 .../com/google/gerrit/reviewdb/client/Project.java |   81 +++++++++--
 .../java/com/google/gerrit/server/UrlEncoded.java  |   10 +-
 .../google/gerrit/server/git/ProjectConfig.java    |    8 +-
 .../gerrit/server/project/DashboardResource.java   |   23 ++-
 .../server/project/DashboardsCollection.java       |   70 +++++----
 .../gerrit/server/project/DeleteDashboard.java     |   12 +-
 .../google/gerrit/server/project/GetDashboard.java |   32 ++--
 .../gerrit/server/project/ListDashboards.java      |   20 ++--
 .../google/gerrit/server/project/SetDashboard.java |   10 +-
 .../gerrit/server/project/SetDefaultDashboard.java |  156 ------------------
 .../gerrit/server/project/SetTypedDashboard.java   |  165 ++++++++++++++++++++
 18 files changed, 392 insertions(+), 254 deletions(-)
 delete mode 100644 gerrit-server/src/main/java/com/google/gerrit/server/project/SetDefaultDashboard.java
 create mode 100644 gerrit-server/src/main/java/com/google/gerrit/server/project/SetTypedDashboard.java

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
index d7db9ac..b813932 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
@@ -407,6 +407,7 @@ public class Dispatcher {
       rest = rest.substring(c);
       if (matchPrefix(DASHBOARDS, rest)) {
         final String dashboardId = skip(rest);
+        final Project.DashboardType type = Project.DashboardType.fromId(dashboardId);
         GerritCallback<DashboardInfo> cb = new GerritCallback<DashboardInfo>() {
           @Override
           public void onSuccess(DashboardInfo result) {
@@ -420,7 +421,7 @@ public class Dispatcher {
 
           @Override
           public void onFailure(Throwable caught) {
-            if ("default".equals(dashboardId) && RestApi.isNotFound(caught)) {
+            if (Project.DashboardType.DEFAULT == type && RestApi.isNotFound(caught)) {
               Gerrit.display(PageLinks.toChangeQuery(
                   PageLinks.projectQuery(new Project.NameKey(project))));
             } else {
@@ -428,8 +429,8 @@ public class Dispatcher {
             }
           }
         };
-        if ("default".equals(dashboardId)) {
-          DashboardList.getDefault(new Project.NameKey(project), cb);
+        if (type != null) {
+          DashboardList.getType(new Project.NameKey(project), type, cb);
           return;
         }
         c = dashboardId.indexOf(":");
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.java
index d83bad9..d84df86 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.java
@@ -21,5 +21,4 @@ public interface DashboardConstants extends Constants {
   String dashboardDescription();
   String dashboardInherited();
   String dashboardItem();
-  String dashboardDefaultToolTip();
 }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.properties
index 4e999b3..eb5fa43 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardConstants.properties
@@ -2,4 +2,3 @@ dashboardName = Dashboard Name
 dashboardDescription = Dashboard Description
 dashboardInherited = Inherited From
 dashboardItem = dashboard
-dashboardDefaultToolTip = Project Default Dashboard
\ No newline at end of file
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardInfo.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardInfo.java
index 9eb42f3..57f6128 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardInfo.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardInfo.java
@@ -14,8 +14,13 @@
 
 package com.google.gerrit.client.dashboards;
 
+import com.google.gerrit.client.rpc.NativeList;
+import com.google.gerrit.reviewdb.client.Project;
 import com.google.gwt.core.client.JavaScriptObject;
 
+import java.util.HashSet;
+import java.util.Set;
+
 public class DashboardInfo extends JavaScriptObject {
   public final native String id() /*-{ return this.id; }-*/;
   public final native String project() /*-{ return this.project; }-*/;
@@ -25,8 +30,21 @@ public class DashboardInfo extends JavaScriptObject {
   public final native String description() /*-{ return this.description; }-*/;
   public final native String foreach() /*-{ return this.foreach; }-*/;
   public final native String url() /*-{ return this.url; }-*/;
-  public final native boolean isDefault() /*-{ return this['default'] ? true : false; }-*/;
+  private final native NativeList<DashboardType> types0() /*-{ return this.types; }-*/;
 
   protected DashboardInfo() {
   }
+
+  public static class DashboardType extends JavaScriptObject {
+    protected DashboardType() {
+    }
+  }
+
+  public final Set<Project.DashboardType> types() {
+    Set<Project.DashboardType> dts = new HashSet<Project.DashboardType>();
+    for (DashboardType t : types0().asList()) {
+      dts.add(Project.DashboardType.valueOf(t.toString()));
+    }
+    return dts;
+  }
 }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardList.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardList.java
index f4da5fa..0558858 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardList.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardList.java
@@ -29,9 +29,9 @@ public class DashboardList extends NativeList<DashboardInfo> {
       .get(callback);
   }
 
-  public static void getDefault(Project.NameKey project,
+  public static void getType(Project.NameKey project, Project.DashboardType type,
       AsyncCallback<DashboardInfo> callback) {
-    new RestApi(base(project) + "default")
+    new RestApi(base(project) + type.id)
       .addParameterTrue("inherited")
       .get(callback);
   }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardsTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardsTable.java
index 724307e..e7e0210 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardsTable.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/dashboards/DashboardsTable.java
@@ -22,14 +22,20 @@ import com.google.gerrit.reviewdb.client.Project;
 import com.google.gwt.user.client.History;
 import com.google.gwt.user.client.ui.Anchor;
 import com.google.gwt.user.client.ui.FlexTable.FlexCellFormatter;
+import com.google.gwt.user.client.ui.FlowPanel;
 import com.google.gwt.user.client.ui.Image;
+import com.google.gwt.user.client.ui.Widget;
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+
+import com.google.gwt.user.client.ui.Label;
 
 public class DashboardsTable extends NavigationTable<DashboardInfo> {
   Project.NameKey project;
@@ -117,11 +123,7 @@ public class DashboardsTable extends NavigationTable<DashboardInfo> {
   }
 
   protected void populate(final int row, final DashboardInfo k) {
-    if (k.isDefault()) {
-      table.setWidget(row, 1, new Image(Gerrit.RESOURCES.greenCheck()));
-      final FlexCellFormatter fmt = table.getFlexCellFormatter();
-      fmt.getElement(row, 1).setTitle(Util.C.dashboardDefaultToolTip());
-    }
+    table.setWidget(row, 1, createTypesWidget(k.types()));
     table.setWidget(row, 2, new Anchor(k.path(), "#"
             + PageLinks.toProjectDashboard(new Project.NameKey(k.project()), k.id())));
     table.setText(row, 3, k.description());
@@ -132,6 +134,16 @@ public class DashboardsTable extends NavigationTable<DashboardInfo> {
     setRowItem(row, k);
   }
 
+  private Widget createTypesWidget(Collection<Project.DashboardType> types) {
+    FlowPanel fp = new FlowPanel();
+    for (Project.DashboardType type : types) {
+      Label l = new Label(type.toString().substring(0, 1));
+      l.setTitle(type.toString());
+      fp.add(l);
+    }
+    return fp;
+  }
+
   @Override
   protected Object getRowItemKey(final DashboardInfo item) {
     return item.id();
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
index 5418be3..6357e14 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
@@ -43,6 +43,9 @@ import com.google.gerrit.server.git.GitRepositoryManager;
 import com.google.gerrit.server.git.MergeOp;
 import com.google.gerrit.server.patch.PatchSetInfoNotAvailableException;
 import com.google.gerrit.server.project.ChangeControl;
+import com.google.gerrit.server.project.GetDashboard;
+import com.google.gerrit.server.project.DashboardResource;
+import com.google.gerrit.server.project.DashboardsCollection.DashboardInfo;
 import com.google.gerrit.server.project.NoSuchChangeException;
 import com.google.gerrit.server.query.change.ChangeData;
 import com.google.gerrit.server.workflow.CategoryFunction;
@@ -50,6 +53,7 @@ import com.google.gerrit.server.workflow.FunctionState;
 import com.google.gwtorm.server.OrmException;
 import com.google.gwtorm.server.ResultSet;
 import com.google.inject.Inject;
+import com.google.inject.Provider;
 import com.google.inject.assistedinject.Assisted;
 
 import org.eclipse.jgit.errors.RepositoryNotFoundException;
diff --git a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Project.java b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Project.java
index c070e3e..987a9ce 100644
--- a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Project.java
+++ b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Project.java
@@ -17,6 +17,14 @@ package com.google.gerrit.reviewdb.client;
 import com.google.gwtorm.client.Column;
 import com.google.gwtorm.client.StringKey;
 
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
 /** Projects match a source code repository managed by Gerrit */
 public final class Project {
   /** Project name key */
@@ -91,6 +99,36 @@ public final class Project {
     INHERIT;
   }
 
+  public static enum DashboardType {
+    DEFAULT("default");
+
+    public final String id;
+
+    DashboardType(String id) {
+      this.id = id;
+    }
+
+    public static DashboardType fromId(String id) {
+      for (DashboardType type: values()) {
+        if (type.id.equals(id)) {
+          return type;
+        }
+      }
+      return null;
+    }
+
+    public static Set<Project.DashboardType> asSet() {
+      Set<Project.DashboardType> types = new HashSet<Project.DashboardType>();
+      Collections.addAll(types, values());
+      return types;
+    }
+
+    public static List<Project.DashboardType> asList() {
+      return Arrays.asList(values());
+    }
+  }
+
+
   protected NameKey name;
 
   protected String description;
@@ -109,9 +147,9 @@ public final class Project {
 
   protected InheritableBoolean useContentMerge;
 
-  protected String defaultDashboardId;
+  protected Map<DashboardType,String> dashboardIdByType;
 
-  protected String localDefaultDashboardId;
+  protected Map<DashboardType,String> localDashboardIdByType;
 
   protected Project() {
   }
@@ -190,20 +228,43 @@ public final class Project {
     state = newState;
   }
 
-  public String getDefaultDashboard() {
-    return defaultDashboardId;
+  public void setDashboard(final DashboardType type, final String dashboardId) {
+    if (dashboardIdByType == null) {
+      dashboardIdByType = new HashMap<DashboardType,String>();
+    }
+    if (dashboardId == null) {
+      dashboardIdByType.remove(type);
+      return;
+    }
+    dashboardIdByType.put(type, dashboardId);
   }
 
-  public void setDefaultDashboard(final String defaultDashboardId) {
-    this.defaultDashboardId = defaultDashboardId;
+  public String getDashboard(DashboardType type) {
+    return dashboardIdByType != null ? dashboardIdByType.get(type) : null;
+  }
+
+  public Map<DashboardType,String> getDashboardIdByType() {
+    return dashboardIdByType;
+  }
+
+  public void setLocalDashboard(final DashboardType type,
+      final String dashboardId) {
+    if (localDashboardIdByType == null) {
+      localDashboardIdByType = new HashMap<DashboardType,String>();
+    }
+    if (dashboardId == null) {
+      localDashboardIdByType.remove(type);
+      return;
+    }
+    localDashboardIdByType.put(type, dashboardId);
   }
 
-  public String getLocalDefaultDashboard() {
-    return localDefaultDashboardId;
+  public String getLocalDashboard(DashboardType type) {
+    return localDashboardIdByType != null ? localDashboardIdByType.get(type) : null;
   }
 
-  public void setLocalDefaultDashboard(final String localDefaultDashboardId) {
-    this.localDefaultDashboardId = localDefaultDashboardId;
+  public Map<DashboardType,String> getLocalDashboardIdByType() {
+    return localDashboardIdByType;
   }
 
   public void copySettingsFrom(final Project update) {
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/UrlEncoded.java b/gerrit-server/src/main/java/com/google/gerrit/server/UrlEncoded.java
index fc27fa1..3a12201 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/UrlEncoded.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/UrlEncoded.java
@@ -35,11 +35,17 @@ public class UrlEncoded extends LinkedHashMap<String, String> {
   @Override
   public String toString() {
     final StringBuilder buffer = new StringBuilder();
-    char separator = 0;
     if (url != null) {
-      separator = '?';
       buffer.append(url);
+      buffer.append("?");
     }
+    buffer.append(getParameters());
+    return buffer.toString();
+  }
+
+  public String getParameters() {
+    final StringBuilder buffer = new StringBuilder();
+    char separator = 0;
     for (final Map.Entry<String, String> entry : entrySet()) {
       final String key = entry.getKey();
       final String val = entry.getValue();
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/git/ProjectConfig.java b/gerrit-server/src/main/java/com/google/gerrit/server/git/ProjectConfig.java
index 1c1c403..c950517 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/git/ProjectConfig.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/git/ProjectConfig.java
@@ -295,8 +295,8 @@ public class ProjectConfig extends VersionedMetaData {
     p.setUseContentMerge(getEnum(rc, SUBMIT, null, KEY_MERGE_CONTENT, Project.InheritableBoolean.INHERIT));
     p.setState(getEnum(rc, PROJECT, null, KEY_STATE, defaultStateValue));
 
-    p.setDefaultDashboard(rc.getString(DASHBOARD, null, KEY_DEFAULT));
-    p.setLocalDefaultDashboard(rc.getString(DASHBOARD, null, KEY_LOCAL_DEFAULT));
+    p.setDashboard(Project.DashboardType.DEFAULT, rc.getString(DASHBOARD, null, KEY_DEFAULT));
+    p.setLocalDashboard(Project.DashboardType.DEFAULT, rc.getString(DASHBOARD, null, KEY_LOCAL_DEFAULT));
 
     loadAccountsSection(rc, groupsByName);
     loadContributorAgreements(rc, groupsByName);
@@ -545,8 +545,8 @@ public class ProjectConfig extends VersionedMetaData {
 
     set(rc, PROJECT, null, KEY_STATE, p.getState(), null);
 
-    set(rc, DASHBOARD, null, KEY_DEFAULT, p.getDefaultDashboard());
-    set(rc, DASHBOARD, null, KEY_LOCAL_DEFAULT, p.getLocalDefaultDashboard());
+    set(rc, DASHBOARD, null, KEY_DEFAULT, p.getDashboard(Project.DashboardType.DEFAULT));
+    set(rc, DASHBOARD, null, KEY_LOCAL_DEFAULT, p.getLocalDashboard(Project.DashboardType.DEFAULT));
 
     Set<AccountGroup.UUID> keepGroups = new HashSet<AccountGroup.UUID>();
     saveAccountsSection(rc, keepGroups);
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java
index 0e2db48..4672ff5 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java
@@ -16,6 +16,7 @@ package com.google.gerrit.server.project;
 
 import com.google.gerrit.extensions.restapi.RestResource;
 import com.google.gerrit.extensions.restapi.RestView;
+import com.google.gerrit.reviewdb.client.Project;
 import com.google.inject.TypeLiteral;
 
 import org.eclipse.jgit.lib.Config;
@@ -24,26 +25,34 @@ public class DashboardResource implements RestResource {
   public static final TypeLiteral<RestView<DashboardResource>> DASHBOARD_KIND =
       new TypeLiteral<RestView<DashboardResource>>() {};
 
-  static DashboardResource projectDefault(ProjectControl ctl) {
-    return new DashboardResource(ctl, null, null, null, true);
+  public static DashboardResource projectTyped(ProjectControl ctl,
+      Project.DashboardType type) {
+    return new DashboardResource(ctl, null, null, null, type);
   }
 
   private final ProjectControl control;
   private final String refName;
   private final String pathName;
   private final Config config;
-  private final boolean projectDefault;
+  private Project.DashboardType type;
+
+  DashboardResource(ProjectControl control,
+      String refName,
+      String pathName,
+      Config config) {
+    this(control, refName, pathName, config, null);
+  }
 
   DashboardResource(ProjectControl control,
       String refName,
       String pathName,
       Config config,
-      boolean projectDefault) {
+      Project.DashboardType type) {
     this.control = control;
     this.refName = refName;
     this.pathName = pathName;
     this.config = config;
-    this.projectDefault = projectDefault;
+    this.type = type;
   }
 
   public ProjectControl getControl() {
@@ -62,7 +71,7 @@ public class DashboardResource implements RestResource {
     return config;
   }
 
-  public boolean isProjectDefault() {
-    return projectDefault;
+  public Project.DashboardType getType() {
+    return type;
   }
 }
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java
index 29544ca..fa7faee 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java
@@ -48,26 +48,29 @@ import org.eclipse.jgit.lib.Repository;
 
 import java.io.IOException;
 import java.io.UnsupportedEncodingException;
+import java.util.Collections;
 import java.util.List;
+import java.util.HashSet;
+import java.util.Map;
 import java.util.Set;
 
-class DashboardsCollection implements
+public class DashboardsCollection implements
     ChildCollection<ProjectResource, DashboardResource>,
     AcceptsCreate<ProjectResource>{
   private final GitRepositoryManager gitManager;
   private final DynamicMap<RestView<DashboardResource>> views;
   private final Provider<ListDashboards> list;
-  private final Provider<SetDefaultDashboard.CreateDefault> createDefault;
+  private final Provider<SetTypedDashboard.CreateTyped> createTyped;
 
   @Inject
-  DashboardsCollection(GitRepositoryManager gitManager,
+  public DashboardsCollection(GitRepositoryManager gitManager,
       DynamicMap<RestView<DashboardResource>> views,
       Provider<ListDashboards> list,
-      Provider<SetDefaultDashboard.CreateDefault> createDefault) {
+      Provider<SetTypedDashboard.CreateTyped> createTyped) {
     this.gitManager = gitManager;
     this.views = views;
     this.list = list;
-    this.createDefault = createDefault;
+    this.createTyped = createTyped;
   }
 
   @Override
@@ -79,8 +82,9 @@ class DashboardsCollection implements
   @Override
   public RestModifyView<ProjectResource, ?> create(ProjectResource parent,
       String id) throws RestApiException {
-    if ("default".equals(id)) {
-      return createDefault.get();
+    Project.DashboardType type = Project.DashboardType.fromId(id);
+    if (type != null) {
+      return createTyped.get().setType(type);
     }
     throw new ResourceNotFoundException(id);
   }
@@ -89,8 +93,9 @@ class DashboardsCollection implements
   public DashboardResource parse(ProjectResource parent, String id)
       throws ResourceNotFoundException, IOException, ConfigInvalidException {
     ProjectControl myCtl = parent.getControl();
-    if ("default".equals(id)) {
-      return DashboardResource.projectDefault(myCtl);
+    Project.DashboardType type = Project.DashboardType.fromId(id);
+    if (type != null) {
+      return DashboardResource.projectTyped(myCtl, type);
     }
 
     List<String> parts = Lists.newArrayList(
@@ -146,7 +151,7 @@ class DashboardsCollection implements
         throw new ResourceNotFoundException(id);
       }
       BlobBasedConfig cfg = new BlobBasedConfig(null, git, objId);
-      return new DashboardResource(myCtl, ref, path, cfg, false);
+      return new DashboardResource(myCtl, ref, path, cfg);
     } finally {
       git.close();
     }
@@ -158,7 +163,8 @@ class DashboardsCollection implements
   }
 
   static DashboardInfo parse(Project definingProject, String refName,
-      String path, Config config, String project, boolean setDefault)
+      String path, Config config, String project,
+      Set<Project.DashboardType> setTypes)
       throws UnsupportedEncodingException {
     DashboardInfo info = new DashboardInfo(refName, path);
     info.project = project;
@@ -167,13 +173,15 @@ class DashboardsCollection implements
     info.description = config.getString("dashboard", null, "description");
     info.foreach = config.getString("dashboard", null, "foreach");
 
-    if (setDefault) {
-      String id = refName + ":" + path;
-      info.isDefault = id.equals(defaultOf(definingProject)) ? true : null;
+    String id = refName + ":" + path;
+    for (Project.DashboardType type : types(definingProject, id)) {
+      if (setTypes.contains(type)) {
+        info.types.add(type);
+      }
     }
 
     UrlEncoded u = new UrlEncoded("/dashboard/");
-    u.put("title", Objects.firstNonNull(info.title, info.path));
+    u.put("title", Objects.firstNonNull(info.title, info.path)
     if (info.foreach != null) {
       u.put("foreach", replace(project, info.foreach));
     }
@@ -184,6 +192,7 @@ class DashboardsCollection implements
       u.put(s.name, replace(project, s.query));
       info.sections.add(s);
     }
+    info.parameters = u.getParameters().replace("%3A", ":");
     info.url = u.toString().replace("%3A", ":");
 
     return info;
@@ -193,18 +202,24 @@ class DashboardsCollection implements
     return query.replace("${project}", project);
   }
 
-  private static String defaultOf(Project proj) {
-    final String defaultId = Objects.firstNonNull(
-        proj.getLocalDefaultDashboard(),
-        Strings.nullToEmpty(proj.getDefaultDashboard()));
-    if (defaultId.startsWith(REFS_DASHBOARDS)) {
-      return defaultId.substring(REFS_DASHBOARDS.length());
-    } else {
-      return defaultId;
+  private static Set<Project.DashboardType> types(Project proj, String myId) {
+    Map<Project.DashboardType,String> all = proj.getDashboardIdByType();
+    all.putAll(proj.getLocalDashboardIdByType());
+
+    Set<Project.DashboardType> types = new HashSet<Project.DashboardType>();
+    for (Map.Entry<Project.DashboardType,String> e : all.entrySet()) {
+      String id = e.getValue();
+      if (id.startsWith(REFS_DASHBOARDS)) {
+        id = id.substring(REFS_DASHBOARDS.length());
+      }
+      if (myId.equals(id)) {
+        types.add(e.getKey());
+      }
     }
+    return types;
   }
 
-  static class DashboardInfo {
+  public static class DashboardInfo {
     final String kind = "gerritcodereview#dashboard";
     String id;
     String project;
@@ -213,12 +228,11 @@ class DashboardsCollection implements
     String path;
     String description;
     String foreach;
+    public String parameters;
     String url;
+    List<Project.DashboardType> types = Lists.newArrayList();
 
-    @SerializedName("default")
-    Boolean isDefault;
-
-    String title;
+    public String title;
     List<Section> sections = Lists.newArrayList();
 
     DashboardInfo(String ref, String name)
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/DeleteDashboard.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/DeleteDashboard.java
index 84473f4..97c7393 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/DeleteDashboard.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/DeleteDashboard.java
@@ -22,17 +22,18 @@ import com.google.gerrit.extensions.restapi.RestModifyView;
 import com.google.gerrit.server.project.DeleteDashboard.Input;
 import com.google.inject.Inject;
 import com.google.inject.Provider;
+import com.google.gerrit.reviewdb.client.Project;
 
 class DeleteDashboard implements RestModifyView<DashboardResource, Input> {
   static class Input {
     String commitMessage;
   }
 
-  private final Provider<SetDefaultDashboard> defaultSetter;
+  private final Provider<SetTypedDashboard> typedSetter;
 
   @Inject
-  DeleteDashboard(Provider<SetDefaultDashboard> defaultSetter) {
-    this.defaultSetter = defaultSetter;
+  DeleteDashboard(Provider<SetTypedDashboard> typedSetter) {
+    this.typedSetter = typedSetter;
   }
 
   @Override
@@ -44,10 +45,11 @@ class DeleteDashboard implements RestModifyView<DashboardResource, Input> {
   public Object apply(DashboardResource resource, Input input)
       throws AuthException, BadRequestException, ResourceConflictException,
       Exception {
-    if (resource.isProjectDefault()) {
+    Project.DashboardType type = resource.getType();
+    if (type != null) {
       SetDashboard.Input in = new SetDashboard.Input();
       in.commitMessage = input != null ? input.commitMessage : null;
-      return defaultSetter.get().apply(resource, in);
+      return typedSetter.get().apply(resource, in);
     }
 
     // TODO: Implement delete of dashboards by API.
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java
index 77a4de0..ba3cf5a 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java
@@ -28,31 +28,33 @@ import org.eclipse.jgit.errors.ConfigInvalidException;
 import org.kohsuke.args4j.Option;
 
 import java.io.IOException;
+import java.util.Collections;
 import java.util.Set;
 
-class GetDashboard implements RestReadView<DashboardResource> {
+public class GetDashboard implements RestReadView<DashboardResource> {
   private final DashboardsCollection dashboards;
 
   @Option(name = "--inherited", usage = "include inherited dashboards")
   private boolean inherited;
 
   @Inject
-  GetDashboard(DashboardsCollection dashboards) {
+  public GetDashboard(DashboardsCollection dashboards) {
     this.dashboards = dashboards;
   }
 
   @Override
   public DashboardInfo apply(DashboardResource resource)
       throws ResourceNotFoundException, IOException, ConfigInvalidException {
-    if (inherited && !resource.isProjectDefault()) {
-      // inherited flag can only be used with default.
+    Project.DashboardType type = resource.getType();
+    if (inherited && type == null) {
+      // inherited flag can only be used with types.
       throw new ResourceNotFoundException("inherited");
     }
 
     String project = resource.getControl().getProject().getName();
-    if (resource.isProjectDefault()) {
-      // The default is not resolved to a definition yet.
-      resource = defaultOf(resource.getControl());
+    if (type != null) {
+      // The type is not resolved to a definition yet.
+      resource = typeOf(resource.getControl(), type);
     }
 
     return DashboardsCollection.parse(
@@ -61,16 +63,17 @@ class GetDashboard implements RestReadView<DashboardResource> {
         resource.getPathName(),
         resource.getConfig(),
         project,
-        true);
+        Collections.singleton(type));
   }
 
-  private DashboardResource defaultOf(ProjectControl ctl)
+  private DashboardResource typeOf(ProjectControl ctl, Project.DashboardType type)
       throws ResourceNotFoundException, IOException, ConfigInvalidException {
-    String id = ctl.getProject().getLocalDefaultDashboard();
+    String id = ctl.getProject().getLocalDashboard(type);
     if (Strings.isNullOrEmpty(id)) {
-      id = ctl.getProject().getDefaultDashboard();
+      id = ctl.getProject().getDashboard(type);
     }
-    if ("default".equals(id)) {
+    type = Project.DashboardType.fromId(id);
+    if (type != null) {
       throw new ResourceNotFoundException();
     } else if (!Strings.isNullOrEmpty(id)) {
       return dashboards.parse(new ProjectResource(ctl), id);
@@ -82,8 +85,9 @@ class GetDashboard implements RestReadView<DashboardResource> {
     seen.add(ctl.getProject().getNameKey());
     ProjectState ps = ctl.getProjectState().getParentState();
     while (ps != null && seen.add(ps.getProject().getNameKey())) {
-      id = ps.getProject().getDefaultDashboard();
-      if ("default".equals(id)) {
+      id = ps.getProject().getDashboard(type);
+      type = Project.DashboardType.fromId(id);
+      if (type != null) {
         throw new ResourceNotFoundException();
       } else if (!Strings.isNullOrEmpty(id)) {
         ctl = ps.controlFor(ctl.getCurrentUser());
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java
index 429007d..c5e04ca 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java
@@ -39,6 +39,7 @@ import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.util.List;
+import java.util.HashSet;
 import java.util.Set;
 
 class ListDashboards implements RestReadView<ProjectResource> {
@@ -59,20 +60,18 @@ class ListDashboards implements RestReadView<ProjectResource> {
       throws ResourceNotFoundException, IOException {
   ProjectControl ctl = resource.getControl();
     String project = ctl.getProject().getName();
+    Set<Project.DashboardType> setTypes = Project.DashboardType.asSet();
     if (!inherited) {
-      return scan(resource.getControl(), project, true);
+      return scan(resource.getControl(), project, setTypes);
     }
 
     List<List<DashboardInfo>> all = Lists.newArrayList();
     Set<Project.NameKey> seen = Sets.newHashSet();
-    boolean setDefault = true;
     for (;;) {
       if (ctl.isVisible()) {
-        List<DashboardInfo> list = scan(ctl, project, setDefault);
+        List<DashboardInfo> list = scan(ctl, project, setTypes);
         for (DashboardInfo d : list) {
-          if (d.isDefault != null && Boolean.TRUE.equals(d.isDefault)) {
-            setDefault = false;
-          }
+          setTypes.removeAll(d.types);
         }
         if (!list.isEmpty()) {
           all.add(list);
@@ -89,7 +88,7 @@ class ListDashboards implements RestReadView<ProjectResource> {
   }
 
   private List<DashboardInfo> scan(ProjectControl ctl, String project,
-      boolean setDefault) throws ResourceNotFoundException, IOException {
+      Set<Project.DashboardType> setTypes) throws ResourceNotFoundException, IOException {
     Repository git;
     try {
       git = gitManager.openRepository(ctl.getProject().getNameKey());
@@ -103,7 +102,7 @@ class ListDashboards implements RestReadView<ProjectResource> {
         for (Ref ref : git.getRefDatabase().getRefs(REFS_DASHBOARDS).values()) {
           if (ctl.controlForRef(ref.getName()).canRead()) {
             all.addAll(scanDashboards(ctl.getProject(), git, rw, ref,
-                project, setDefault));
+                project, setTypes));
           }
         }
         return all;
@@ -116,7 +115,8 @@ class ListDashboards implements RestReadView<ProjectResource> {
   }
 
   private List<DashboardInfo> scanDashboards(Project definingProject,
-      Repository git, RevWalk rw, Ref ref, String project, boolean setDefault)
+      Repository git, RevWalk rw, Ref ref, String project,
+      Set<Project.DashboardType> setTypes)
       throws IOException {
     List<DashboardInfo> list = Lists.newArrayList();
     TreeWalk tw = new TreeWalk(rw.getObjectReader());
@@ -132,7 +132,7 @@ class ListDashboards implements RestReadView<ProjectResource> {
                 tw.getPathString(),
                 new BlobBasedConfig(null, git, tw.getObjectId(0)),
                 project,
-                setDefault));
+                setTypes));
           } catch (ConfigInvalidException e) {
             log.warn(String.format(
                 "Cannot parse dashboard %s:%s:%s: %s",
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDashboard.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDashboard.java
index 7c4c886..83e645f 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDashboard.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDashboard.java
@@ -31,11 +31,11 @@ class SetDashboard implements RestModifyView<DashboardResource, Input> {
     String commitMessage;
   }
 
-  private final Provider<SetDefaultDashboard> defaultSetter;
+  private final Provider<SetTypedDashboard> typedSetter;
 
   @Inject
-  SetDashboard(Provider<SetDefaultDashboard> defaultSetter) {
-    this.defaultSetter = defaultSetter;
+  SetDashboard(Provider<SetTypedDashboard> typedSetter) {
+    this.typedSetter = typedSetter;
   }
 
   @Override
@@ -47,8 +47,8 @@ class SetDashboard implements RestModifyView<DashboardResource, Input> {
   public Object apply(DashboardResource resource, Input input)
       throws AuthException, BadRequestException, ResourceConflictException,
       Exception {
-    if (resource.isProjectDefault()) {
-      return defaultSetter.get().apply(resource, input);
+    if (resource.getType() != null) {
+      return typedSetter.get().apply(resource, input);
     }
 
     // TODO: Implement creation/update of dashboards by API.
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDefaultDashboard.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDefaultDashboard.java
deleted file mode 100644
index 7fb2986..0000000
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/SetDefaultDashboard.java
+++ /dev/null
@@ -1,156 +0,0 @@
-// Copyright (C) 2012 The Android Open Source Project
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-// http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-package com.google.gerrit.server.project;
-
-import com.google.common.base.Objects;
-import com.google.common.base.Strings;
-import com.google.gerrit.extensions.restapi.AuthException;
-import com.google.gerrit.extensions.restapi.BadRequestException;
-import com.google.gerrit.extensions.restapi.Response;
-import com.google.gerrit.extensions.restapi.ResourceConflictException;
-import com.google.gerrit.extensions.restapi.ResourceNotFoundException;
-import com.google.gerrit.extensions.restapi.RestModifyView;
-import com.google.gerrit.reviewdb.client.Project;
-import com.google.gerrit.server.IdentifiedUser;
-import com.google.gerrit.server.git.MetaDataUpdate;
-import com.google.gerrit.server.git.ProjectConfig;
-import com.google.gerrit.server.project.DashboardsCollection.DashboardInfo;
-import com.google.gerrit.server.project.SetDashboard.Input;
-import com.google.inject.Inject;
-import com.google.inject.Provider;
-
-import org.eclipse.jgit.errors.ConfigInvalidException;
-import org.eclipse.jgit.errors.RepositoryNotFoundException;
-import org.kohsuke.args4j.Option;
-
-class SetDefaultDashboard implements RestModifyView<DashboardResource, Input> {
-  private final ProjectCache cache;
-  private final MetaDataUpdate.Server updateFactory;
-  private final DashboardsCollection dashboards;
-  private final Provider<GetDashboard> get;
-
-  @Option(name = "--inherited", usage = "set dashboard inherited by children")
-  private boolean inherited;
-
-  @Inject
-  SetDefaultDashboard(ProjectCache cache,
-      MetaDataUpdate.Server updateFactory,
-      DashboardsCollection dashboards,
-      Provider<GetDashboard> get) {
-    this.cache = cache;
-    this.updateFactory = updateFactory;
-    this.dashboards = dashboards;
-    this.get = get;
-  }
-
-  @Override
-  public Class<Input> inputType() {
-    return Input.class;
-  }
-
-  @Override
-  public Object apply(DashboardResource resource, Input input)
-      throws AuthException, BadRequestException, ResourceConflictException,
-      Exception {
-    if (input == null) {
-      input = new Input(); // Delete would set input to null.
-    }
-    input.id = Strings.emptyToNull(input.id);
-
-    ProjectControl ctl = resource.getControl();
-    IdentifiedUser user = (IdentifiedUser) ctl.getCurrentUser();
-    if (!ctl.isOwner()) {
-      throw new AuthException("not project owner");
-    }
-
-    DashboardResource target = null;
-    if (input.id != null) {
-      try {
-        target = dashboards.parse(new ProjectResource(ctl), input.id);
-      } catch (ResourceNotFoundException e) {
-        throw new BadRequestException("dashboard " + input.id + " not found");
-      }
-    }
-
-    try {
-      MetaDataUpdate md = updateFactory.create(ctl.getProject().getNameKey());
-      try {
-        ProjectConfig config = ProjectConfig.read(md);
-        Project project = config.getProject();
-        if (inherited) {
-          project.setDefaultDashboard(input.id);
-        } else {
-          project.setLocalDefaultDashboard(input.id);
-        }
-
-        String msg = Objects.firstNonNull(
-          Strings.emptyToNull(input.commitMessage),
-          input.id == null
-            ? "Removed default dashboard.\n"
-            : String.format("Changed default dashboard to %s.\n", input.id));
-        if (!msg.endsWith("\n")) {
-          msg += "\n";
-        }
-        md.setAuthor(user);
-        md.setMessage(msg);
-        config.commit(md);
-        cache.evict(ctl.getProject());
-
-        if (target != null) {
-          DashboardInfo info = get.get().apply(target);
-          info.isDefault = true;
-          return info;
-        }
-        return Response.none();
-      } finally {
-        md.close();
-      }
-    } catch (RepositoryNotFoundException notFound) {
-      throw new ResourceNotFoundException(ctl.getProject().getName());
-    } catch (ConfigInvalidException e) {
-      throw new ResourceConflictException(String.format(
-          "invalid project.config: %s", e.getMessage()));
-    }
-  }
-
-  static class CreateDefault implements
-      RestModifyView<ProjectResource, SetDashboard.Input> {
-    private final Provider<SetDefaultDashboard> setDefault;
-
-    @Option(name = "--inherited", usage = "set dashboard inherited by children")
-    private boolean inherited;
-
-    @Inject
-    CreateDefault(Provider<SetDefaultDashboard> setDefault) {
-      this.setDefault = setDefault;
-    }
-
-    @Override
-    public Class<Input> inputType() {
-      return SetDashboard.Input.class;
-    }
-
-    @Override
-    public Object apply(ProjectResource resource, Input input)
-        throws AuthException, BadRequestException, ResourceConflictException,
-        Exception {
-      SetDefaultDashboard set = setDefault.get();
-      set.inherited = inherited;
-      return Response.created(set.apply(
-          DashboardResource.projectDefault(resource.getControl()),
-          input));
-    }
-  }
-}
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/SetTypedDashboard.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/SetTypedDashboard.java
new file mode 100644
index 0000000..1c0a541
--- /dev/null
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/SetTypedDashboard.java
@@ -0,0 +1,165 @@
+// Copyright (C) 2012 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.server.project;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Strings;
+import com.google.gerrit.extensions.restapi.AuthException;
+import com.google.gerrit.extensions.restapi.BadRequestException;
+import com.google.gerrit.extensions.restapi.Response;
+import com.google.gerrit.extensions.restapi.ResourceConflictException;
+import com.google.gerrit.extensions.restapi.ResourceNotFoundException;
+import com.google.gerrit.extensions.restapi.RestModifyView;
+import com.google.gerrit.reviewdb.client.Project;
+import com.google.gerrit.server.IdentifiedUser;
+import com.google.gerrit.server.git.MetaDataUpdate;
+import com.google.gerrit.server.git.ProjectConfig;
+import com.google.gerrit.server.project.DashboardsCollection.DashboardInfo;
+import com.google.gerrit.server.project.SetDashboard.Input;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+import org.eclipse.jgit.errors.ConfigInvalidException;
+import org.eclipse.jgit.errors.RepositoryNotFoundException;
+import org.kohsuke.args4j.Option;
+
+import java.util.Set;
+
+class SetTypedDashboard implements RestModifyView<DashboardResource, Input> {
+  private final ProjectCache cache;
+  private final MetaDataUpdate.Server updateFactory;
+  private final DashboardsCollection dashboards;
+  private final Provider<GetDashboard> get;
+
+  @Option(name = "--inherited", usage = "set dashboard inherited by children")
+  private boolean inherited;
+
+  @Inject
+  SetTypedDashboard(ProjectCache cache,
+      MetaDataUpdate.Server updateFactory,
+      DashboardsCollection dashboards,
+      Provider<GetDashboard> get) {
+    this.cache = cache;
+    this.updateFactory = updateFactory;
+    this.dashboards = dashboards;
+    this.get = get;
+  }
+
+  @Override
+  public Class<Input> inputType() {
+    return Input.class;
+  }
+
+  @Override
+  public Object apply(DashboardResource resource, Input input)
+      throws AuthException, BadRequestException, ResourceConflictException,
+      Exception {
+    if (input == null) {
+      input = new Input(); // Delete would set input to null.
+    }
+    input.id = Strings.emptyToNull(input.id);
+
+    ProjectControl ctl = resource.getControl();
+    IdentifiedUser user = (IdentifiedUser) ctl.getCurrentUser();
+    if (!ctl.isOwner()) {
+      throw new AuthException("not project owner");
+    }
+
+    DashboardResource target = null;
+    if (input.id != null) {
+      try {
+        target = dashboards.parse(new ProjectResource(ctl), input.id);
+      } catch (ResourceNotFoundException e) {
+        throw new BadRequestException("dashboard " + input.id + " not found");
+      }
+    }
+
+    try {
+      MetaDataUpdate md = updateFactory.create(ctl.getProject().getNameKey());
+      try {
+        ProjectConfig config = ProjectConfig.read(md);
+        Project project = config.getProject();
+        if (inherited) {
+          project.setDashboard(resource.getType(), input.id);
+        } else {
+          project.setLocalDashboard(resource.getType(), input.id);
+        }
+
+        String type = resource.getType().toString();
+        String msg = Objects.firstNonNull(
+          Strings.emptyToNull(input.commitMessage),
+          input.id == null
+            ? String.format("Removed %s dashboard.\n", type)
+            : String.format("Changed %s dashboard to %s.\n", type, input.id));
+        if (!msg.endsWith("\n")) {
+          msg += "\n";
+        }
+        md.setAuthor(user);
+        md.setMessage(msg);
+        config.commit(md);
+        cache.evict(ctl.getProject());
+
+        if (target != null) {
+          DashboardInfo info = get.get().apply(target);
+          info.types.add(resource.getType());
+          return info;
+        }
+        return Response.none();
+      } finally {
+        md.close();
+      }
+    } catch (RepositoryNotFoundException notFound) {
+      throw new ResourceNotFoundException(ctl.getProject().getName());
+    } catch (ConfigInvalidException e) {
+      throw new ResourceConflictException(String.format(
+          "invalid project.config: %s", e.getMessage()));
+    }
+  }
+
+  static class CreateTyped implements
+      RestModifyView<ProjectResource, SetDashboard.Input> {
+    private final Provider<SetTypedDashboard> setTyped;
+    private Project.DashboardType type;
+
+    @Option(name = "--inherited", usage = "set dashboard inherited by children")
+    private boolean inherited;
+
+    @Inject
+    CreateTyped(Provider<SetTypedDashboard> setTyped) {
+      this.setTyped = setTyped;
+    }
+
+    public CreateTyped setType(Project.DashboardType type) {
+      this.type = type;
+      return this;
+    }
+
+    @Override
+    public Class<Input> inputType() {
+      return SetDashboard.Input.class;
+    }
+
+    @Override
+    public Object apply(ProjectResource resource, Input input)
+        throws AuthException, BadRequestException, ResourceConflictException,
+        Exception {
+      SetTypedDashboard set = setTyped.get();
+      set.inherited = inherited;
+      return Response.created(set.apply(
+          DashboardResource.projectTyped(resource.getControl(), type),
+          input));
+    }
+  }
+}
-- 
1.7.8

